File: Patrones/Delegacion/Java/Delegacion/src/example/Calculator.java
package example;

//changing the implementing object in run-time (normally done in compile time)

//the "delegator"
public class Calculator implements CalculationInterface {
	CalculationInterface ci = null;	
	
	// delegation
	public Calculator(CalculationInterface ci){
		this.setI(ci);
	}
	
	// calls the delegate
	public int add(int a, int b){
		return ci.add(a, b);
	}
	
	public int mul(int a, int b){
		return ci.mul(a, b);
	}
	
	public float div(int a, int b){
		return ci.div(a, b);
	}
	
	public int sub(int a, int b){
		return ci.sub(a, b);
	}
	
	// assign the delegate 
	public void setI(CalculationInterface i){
		this.ci = i;
	}
}

File: Patrones/Delegacion/C++/src/Calculator.cpp
#include <Calculator.h>

Calculator::Calculator(CalculatorInterface * ci){
	this->setI(ci);
}
	
int Calculator::add(int a, int b){
	return this->ci->add(a,b);
}

int Calculator::mul(int a, int b){
	return this->ci->mul(a,b);	
}

float Calculator::div(int a, int b){
	return this->ci->div(a,b);
}

int Calculator::sub(int a, int b){
	return this->ci->sub(a,b);
}

void Calculator::setI(CalculatorInterface * i){
	this->ci = i;
}